From 8deefb326367b65ba66808b774d1ffdd9c0a0a4c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Zolt=C3=A1n=20B=C3=B6sz=C3=B6rm=C3=A9nyi?=
 <zboszor@gmail.com>
Date: Wed, 4 Jun 2025 14:55:08 +0200
Subject: [PATCH] Fix cross-compiling
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

When cross-compiling onnx, the python module is linked as:

[100%] Linking CXX shared module onnx_cpp2py_export.so

instead of

[100%] Linking CXX shared module onnx_cpp2py_export.cpython-313-x86_64-linux-gnu.so

Split the two conflicting usage of find_package(Python3 ...) so
the interpreter and the linker packages can work independently.

See https://gitlab.kitware.com/cmake/cmake/-/issues/25145

Signed-off-by: Zoltán Böszörményi <zboszor@gmail.com>
Upstream-Status: Submitted [https://github.com/onnx/onnx/pull/7026]
---
 CMakeLists.txt | 20 ++++++++++++++------
 1 file changed, 14 insertions(+), 6 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 367fd5ef..1106fe10 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -108,10 +108,6 @@
   endif()
 endif()
 
-if(ONNX_BUILD_PYTHON)
-  set(python_dev_component Development.Module)
-endif()
-
 # explicitly configure FindPython3.cmake to find python3 in virtual environment first
 if(NOT DEFINED Python3_FIND_VIRTUALENV)
   set(Python3_FIND_VIRTUALENV FIRST)
@@ -121,7 +117,22 @@
   set(Python3_FIND_STRATEGY LOCATION)
 endif()
 
-find_package(Python3 REQUIRED COMPONENTS Interpreter ${python_dev_component})
+if(CMAKE_CROSSCOMPILING)
+  # When cross-compiling, the interpreter and the compiling/linking steps
+  # must use a different package. See the discussion about this at
+  # https://gitlab.kitware.com/cmake/cmake/-/issues/25145
+  if(ONNX_BUILD_PYTHON)
+    find_package(Python3 REQUIRED COMPONENTS Development.Module)
+  endif()
+  find_package(Python REQUIRED COMPONENTS Interpreter)
+  set(ONNX_PYTHON_INTERPRETER Python::Interpreter)
+else()
+  if(ONNX_BUILD_PYTHON)
+    set(python_dev_component Development.Module)
+  endif()
+  find_package(Python3 REQUIRED COMPONENTS Interpreter ${python_dev_component})
+  set(ONNX_PYTHON_INTERPRETER Python3::Interpreter)
+endif()
 
 if(CMAKE_SYSTEM_NAME STREQUAL "AIX")
   set(CMAKE_NO_SYSTEM_FROM_IMPORTED 1)
@@ -382,7 +393,7 @@
     endif()
 
     add_custom_command(OUTPUT "${GENERATED_PROTO}"
-                       COMMAND Python3::Interpreter "${GEN_PROTO_PY}"
+                       COMMAND ${ONNX_PYTHON_INTERPRETER} "${GEN_PROTO_PY}"
                                ARGS ${GEN_PROTO_ARGS}
                        DEPENDS ${INFILE}
                        COMMENT "Running gen_proto.py on ${INFILE}"
-- 
2.49.0

